contributors/design-proposals/kubelet-rootfs-distribution.md

```diff
@@ -0,0 +1,199 @@
+
+Currently, the Kubelet does not need access to arbitrary paths on the host (as
+hostPath volumes are managed entirely by the docker daemon process, including
+SELinux context applying), so Kubelet makes no operations at those paths).
+
+This will likely change in the future, at which point a shared bindmount of `/`
+will be made available at a known path in the Kubelet chroot. This change will
+necessarily be more intrusive since it will require the kubelet to behave
+differently (use the shared rootfs mount's path) when running within the
+chroot.
+
+## Current Use
+
+This method of running the Kubelet is already in use by users of CoreOS Linux. The details of this implementation are found in the [kubelet wrapper documentation](https://coreos.com/kubernetes/docs/latest/kubelet-wrapper.html).
+
+## Implementation
+
+### Target Distros
+
+The two distros which benefit the most from this change are GCI and CoreOS. Initially, these changes will only be implemented for those distros.
+
+This work will also only initially target the GCE provider and `kube-up` method of deployment.
+
+#### Hyperkube Image Packaging
+
+The Hyperkube image is distributed as part of an official release to the `gcr.io/google_containers` registry, but is not included along with the `kube-up` artifacts used for deployment.
+
+This will need to be remediated in order to complete this proposal.
+
+### Testing & Rollout
+
+In order to ensure the paths remain complete, e2e tests *must* be run against a
+Kubelet operating in this manner as part of the submit queue.
+
+To ensure that this feature does not unduly impact others, it will be added to
+GCI, but gated behind a feature-flag for a sort confidence-building period
+(e.g.  `KUBE_RUN_HYPERKUBE_IMAGE=false`). A temporary non-blocking e2e job will
+be added with that option. If the results look clean after a week, the
+deployment option can be removed and the GCI image can completely switch over.
+
+Once that testing is in place, it can be rolled out across other distros as
+desired.
+
+
+#### Everything else
+
+In the initial implementation, rkt or docker can be used to extract the rootfs of the hyperkube image. rkt fly or a systemd unit (using [`RootDirectory`](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RootDirectory=)) can be used to perform the needed setup, chroot, and execution of the kubelet within that rootfs.
+
+
+
+## FAQ
+
+#### Will this replace or break other installation options?
+
+Other installation options include using RPMs, DEBs, and simply running the statically compiled Kubelet binary.
+
+All of these methods will continue working as they do now. In the future they may choose to also run the kubelet in this manner, but they don't necessarily have to.
+
+
+#### Is this running the kubelet as a pod?
+
+This is different than running the Kubelet as a pod. Rather than using namespaces, it uses only a chroot and shared bind mounts.
+
+## Alternatives
```

Comment:
Another option/enhancement is to set up some simple resource cgroups so that we can ensure that we can be sure that the kubelet can keep running.  On a heavily loaded system it would be nice to carve out some guaranteed elbow room for the kubelet.  Full containerization (namespace+cgroup) wouldn't be necessary.